### PR TITLE
Add mypy typing to precision plugins.

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?
This PR should be merged after #6148 

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 1]
```diff
@@ -149,7 +152,8 @@ def reinit_scheduler_properties(optimizers: list, schedulers: list):
                     break
 
     def pre_optimizer_step(
-        self, pl_module: LightningModule, optimizer: Optimizer, optimizer_idx: int, lambda_closure: Callable, **kwargs
+        self, pl_module: LightningModule, optimizer: 'Optimizer', optimizer_idx: int, lambda_closure: Callable,
+        **kwargs: Any
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/apex_amp.py:156]:
```suggestion
        **kwargs: Any,
```
 and use yapf :]

File: pytorch_lightning/plugins/precision/deepspeed_precision.py [hunk 2]
```diff
@@ -33,28 +37,30 @@ def pre_optimizer_step(
 
     def backward(
         self,
-        lightning_module: LightningModule,
+        model: 'LightningModule',
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/deepspeed_precision.py:40]: isn't this API change?

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 3]
```diff
@@ -39,16 +42,16 @@ def master_params(self, optimizer: Optimizer) -> Generator[torch.Tensor, None, N
             for p in group["params"]:
                 yield p
 
-    def connect(self, model: Module, optimizers: Sequence,
-                lr_schedulers: Sequence) -> Tuple[Module, Sequence, Sequence]:
+    def connect(self, model: 'Module', optimizers: Sequence['Optimizer'],
+                lr_schedulers: Sequence[Any]) -> Tuple['Module', Sequence['Optimizer'], Sequence[Any]]:
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/precision_plugin.py:46]:
```suggestion
                lr_schedulers: Sequence[Any],) -> Tuple['Module', Sequence['Optimizer'], Sequence[Any]]:
```
plus yapf :]

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 4]
```diff
@@ -78,15 +81,18 @@ def backward(
         return closure_loss
 
     def pre_optimizer_step(
-        self, pl_module: LightningModule, optimizer: Optimizer, optimizer_idx: int, closure: Callable, **kwargs
+        self, pl_module: 'LightningModule', optimizer: 'Optimizer', optimizer_idx: int, lambda_closure: Callable,
+        **kwargs: Any
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/precision_plugin.py:85]:
```suggestion
        **kwargs: Any,
```

File: pytorch_lightning/plugins/precision/tpu_bfloat.py [hunk 5]
```diff
@@ -12,17 +12,21 @@
...
 
 class TPUHalfPrecisionPlugin(PrecisionPlugin):
     """Plugin that enables bfloats on TPUs"""
 
-    precision = 16
+    precision: int = 16
 
-    def connect(self, model: torch.nn.Module, optimizers, lr_schedulers):
+    def connect(self, model: 'Module', optimizers: Sequence['Optimizer'],
+                lr_schedulers: Sequence[Any]) -> Tuple[Module, Sequence['Optimizer'], Sequence[Any]]:
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/tpu_bfloat.py:30]:
```suggestion
                lr_schedulers: Sequence[Any],) -> Tuple[Module, Sequence['Optimizer'], Sequence[Any]]:
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 6]
```diff
@@ -35,6 +35,11 @@ def __init__(self) -> None:
         self._results = None
         self.global_rank = 0
 
+    @abstractmethod
+    def connect(self, model: 'Module') -> None:
+        """Called by the accelerator to connect it with this plugin
+        """
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:41]:
```suggestion
        """Called by the accelerator to connect it with this plugin"""
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/plugins/precision/deepspeed_precision.py [hunk 7]
```diff
@@ -1,24 +1,28 @@
...
 
-    def __init__(self, precision):
+    def __init__(self, precision: int) -> None:
         super().__init__()
         self.precision = precision
 
     def pre_optimizer_step(
-        self, pl_module: LightningModule, optimizer: Optimizer, optimizer_idx: int, lambda_closure: Callable, **kwargs
+        self, pl_module: 'LightningModule', optimizer: 'Optimizer', optimizer_idx: int, lambda_closure: Callable,
+        **kwargs: Any
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/deepspeed_precision.py:25]:
```suggestion
        **kwargs: Any,
```

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 8]
```diff
@@ -65,7 +64,8 @@ def backward(
         return closure_loss
 
     def pre_optimizer_step(
-        self, pl_module: LightningModule, optimizer: Optimizer, optimizer_idx: int, lambda_closure: Callable, **kwargs
+        self, pl_module: 'LightningModule', optimizer: 'Optimizer', optimizer_idx: int, lambda_closure: Callable,
+        **kwargs: Any
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/native_amp.py:68]:
```suggestion
        **kwargs: Any,
```

File: pytorch_lightning/plugins/precision/deepspeed_precision.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/deepspeed_precision.py:44]: It was named model in all the other plugins and I checked, It is never called by name. 
But if we don't do this, MypY complains, that the signature does not match with base class since it has different parameters

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 9]
```diff
@@ -78,15 +85,22 @@ def backward(
...
         """Hook to do something before each optimizer step."""
         return True
 
-    def post_optimizer_step(self, optimizer: Optimizer, optimizer_idx: int) -> None:
+    def post_optimizer_step(self, optimizer: 'Optimizer', optimizer_idx: int) -> None:
         """Hook to do something after each optimizer step."""
 
-    def clip_gradients(self, optimizer: Optimizer, clip_val: Union[int, float], norm_type: float = float(2.0)) -> None:
+    def clip_gradients(
+        self, optimizer: 'Optimizer', clip_val: Union[int, float], norm_type: float = float(2.0)
```
Comment (akihironitta) [inline pytorch_lightning/plugins/precision/precision_plugin.py:102]: Same as above comment
```suggestion
        self, optimizer: 'Optimizer', clip_val: Union[int, float], norm_type: float = 2.0
```

File: pytorch_lightning/plugins/precision/sharded_native_amp.py [hunk 10]
```diff
@@ -22,15 +20,20 @@
...
     """
 
-    def __init__(self):
+    def __init__(self) -> None:
         super().__init__()
         self.scaler = ShardedGradScaler()
 
-    def clip_gradients(self, optimizer: Optimizer, clip_val: Union[int, float], norm_type: float = float(2.0)):
+    def clip_gradients(
+        self, optimizer: 'Optimizer', clip_val: Union[int, float], norm_type: float = float(2.0)
```
Comment (akihironitta) [inline pytorch_lightning/plugins/precision/sharded_native_amp.py:36]: Same as above
```suggestion
        self, optimizer: 'Optimizer', clip_val: Union[int, float], norm_type: float = 2.0
```

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 11]
```diff
@@ -12,37 +12,36 @@
...
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
-if _NATIVE_AMP_AVAILABLE:
-    from torch.cuda.amp import autocast
-else:
-    autocast = None
+if TYPE_CHECKING:
+    from torch.optim import Optimizer
+
+    from pytorch_lightning.core import LightningModule
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:27]: I believe it's to avoid circular imports

[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/plugins/precision/native_amp.py:27]: Yes, the more specific you are with your import the less likely are circular imports due to something being indirectly imported in the __init__.py already!

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 12]
```diff
@@ -12,37 +12,36 @@
...
 
 
 class NativeMixedPrecisionPlugin(MixedPrecisionPlugin):
 
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/native_amp.py:32]: why is it necessary to annotate init with None?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 12 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:32]: https://github.com/python/mypy/issues/3358#issuecomment-301160321

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (akihironitta) [reply pytorch_lightning/plugins/precision/native_amp.py:27]: @carmocca @justusschock That totally makes sense! Thanks :]

**Review by akihironitta — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
